pkg/cloud/rgraph/exec/executor_parallel.go

```diff
@@ -0,0 +1,175 @@
+*/
+
+package exec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph/algo"
+	"k8s.io/klog/v2"
+)
+
+var (
+	ErrPendingActions = errors.New("Executor did not process all actions")
+)
+
+func defaultParallelExecutorConfig() *ExecutorConfig {
+	return &ExecutorConfig{
+		DryRun:        false,
+		ErrorStrategy: ContinueOnError,
+		Timeout:       5 * time.Minute,
+	}
+}
+
+// NewParallelExecutor returns a new Executor that runs tasks multi`-threaded.
+func NewParallelExecutor(pending []Action, opts ...Option) (*parallelExecutor, error) {
+	ret := &parallelExecutor{
+		config: defaultParallelExecutorConfig(),
+		result: &Result{Pending: pending},
+		pq:     algo.NewParallelQueue[Action](),
+	}
+	for _, opt := range opts {
+		opt(ret.config)
+	}
+
+	if err := ret.config.validate(); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+type parallelExecutor struct {
+	config *ExecutorConfig
+	c      cloud.Cloud
+
+	// lock guards results
+	lock   sync.Mutex
+	result *Result
+
+	pq   *algo.ParallelQueue[Action]
+	done chan *TraceEntry
+}
+
+// parallelExecutor implements Executor.
+var _ Executor = (*parallelExecutor)(nil)
+
+func (ex *parallelExecutor) Run(ctx context.Context, c cloud.Cloud) (*Result, error) {
+	ex.queueRunnableActions()
+	subctx, cancel := context.WithTimeout(ctx, ex.config.Timeout)
+
+	runAction := func(ctx context.Context, a Action) error {
+		te := &TraceEntry{
+			Action: a,
+			Start:  time.Now(),
+		}
+		klog.V(4).Infof("Run action %s", a)
+		events, runErr := a.Run(ctx, c)
+		te.End = time.Now()
+		klog.V(4).Infof("Finish action %s, err: %v", a, runErr)
+
+		ex.addActionResult(a, runErr)
+
+		if runErr != nil {
+			klog.Infof("Got error  %v, from action %s error_strategy: %s", runErr, a, ex.config.ErrorStrategy)
+			// check error strategy and decide if new actions should be executed.
+			if ex.config.ErrorStrategy == StopOnError {
+				return fmt.Errorf("parallelExecutor: stopping execution for Action %s (got %w)", a, runErr)
+			}
+		} else {
+			// notify parents only when action finished with success
+			for _, ev := range events {
+				signaled := ex.signal(ev)
+				te.Signaled = append(te.Signaled, signaled...)
+			}
+		}
+
+		if ex.config.Tracer != nil {
+			ex.config.Tracer.Record(te, runErr)
+		}
+
+		// try to run pending tasks
+		ex.queueRunnableActions()
+		return nil
+	}
+
+	queueErr := ex.pq.Run(subctx, runAction)
+	cancel()
+	if queueErr != nil {
+		waitErr := ex.pq.WaitForOrphans(ctx)
+		if waitErr != nil {
+			return ex.result, fmt.Errorf("ParallelExecutor: WaitForOrphans: %w", waitErr)
+		}
+	}
+	if len(ex.result.Errors) > 0 || len(ex.result.Pending) != 0 {
+		return ex.result, ErrPendingActions
+	}
+	return ex.result, nil
+
+}
+
+func (ex *parallelExecutor) queueRunnableActions() {
+	klog.V(4).Infof("queueRunnableActions() pending: %d", len(ex.result.Pending))
```

Comment:
Let makes this easier to understand:

klog.V(4).Infof("queueRunnableActions: %d actions pending", len...)

Reply:
done